File: tree/102.binary-tree-level-order-traversal.py
from typeing import List

#
# @lc app=leetcode id=102 lang=python3
#
# [102] Binary Tree Level Order Traversal
#

# @lc code=start
# Definition for a binary tree node.


class TreeNode:
    def __init__(
        self, val: int = 0, left: "TreeNode" = None, right: "TreeNode" = None
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        queue = []
        result = []
        queue.append(root)
        while queue:
            size = len(queue)
            values = []
            while size > 0:
                node = queue.pop(0)
                values.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
                size -= 1
            result.append(values)
        return result


# @lc code=end

File: tree/102.binary-tree-level-order-traversal.go
package main

/*
 * @lc app=leetcode id=102 lang=golang
 *
 * [102] Binary Tree Level Order Traversal
 */

// @lc code=start
// *
// Definition for a binary tree node.
// type TreeNode struct {
// 	Val   int
// 	Left  *TreeNode
// 	Right *TreeNode
// }

func levelOrder(root *TreeNode) [][]int {
	res := [][]int{}
	dfs(root, 0, &res)
	return res
}

func dfs(node *TreeNode, length int, res *[][]int) {
	if node == nil {
		return
	}
	for len(*res) <= length {
		*res = append(*res, []int{})
	}
	(*res)[length] = append((*res)[length], node.Val)
	dfs(node.Left, length+1, res)
	dfs(node.Right, length+1, res)
}

// @lc code=end
